torchdynamo/_eval_frame.c

```diff
@@ -132,29 +108,29 @@ inline static PyObject *eval_frame_default(PyThreadState *tstate,
 #endif
 }
 
-inline static void enable_eval_frame(PyThreadState *tstate) {
+inline static void enable_eval_frame_shim(PyThreadState *tstate) {
 #if PY_VERSION_HEX >= 0x03090000
-  _PyInterpreterState_SetEvalFrameFunc(tstate->interp, &custom_eval_frame);
-#else
-  tstate->interp->eval_frame = &custom_eval_frame;
-#endif
-}
-
-inline static void disable_eval_frame(PyThreadState *tstate) {
-#if PY_VERSION_HEX >= 0x03090000
-  _PyInterpreterState_SetEvalFrameFunc(tstate->interp,
-                                       &_PyEval_EvalFrameDefault);
+  if (_PyInterpreterState_GetEvalFrameFunc(tstate->interp) != &custom_eval_frame_shim) {
+    _PyInterpreterState_SetEvalFrameFunc(tstate->interp, &custom_eval_frame_shim);
+  }
 #else
-  tstate->interp->eval_frame = &_PyEval_EvalFrameDefault;
+  if (tstate->interp->eval_frame != &custom_eval_frame_shim) { 
```

Comment:
Is this check just an optimization?  Curious about the motivation.

Reply:
Yes, just an optimization. We can always drop it if you think it doesn't add value.